### ISSUE TITLE
Class 'Foo' is compiled by a new Kotlin compiler backend and cannot be loaded by the old compiler

### ISSUE THREAD

**Issue by oh-Rin** [gradle/gradle#15335]
> Gradle Version: `gradle-6.8-rc-1-bin`

I have a plugin compiled with `Kotlin 1.4.20` and referenced it in `settings.gradle.kts`, and then this error occurs. How can I solve it?

![image](https://user-images.githubusercontent.com/58068445/100407841-758ee180-30a4-11eb-8f45-e3527871c784.png)

**Comment by eskatos** [comment]
Hi @oh-Rin,

Does it works from the command line? Is it an IDE-only problem?

In Gradle 6.8 the embedded Kotlin has been updated to 1.4.20 but build scripts still target Kotlin language 1.3 and stdlib 1.3 for backwards compatibility reasons. This will change in 7.0, see https://github.com/gradle/gradle/issues/14888

**Comment by oh-Rin** [comment]
> Hi @oh-Rin,
> 
> Does it works from the command line? Is it an IDE-only problem?
> 
> In Gradle 6.8 the embedded Kotlin has been updated to 1.4.20 but build scripts still target Kotlin language 1.3 and stdlib 1.3 for backwards compatibility reasons. This will change in 7.0, see #14888

Yes, it's just IDE problem, in fact it will not report any errors when running.

Is there way to temporarily suppress this error? 
...You know, this is a bit annoying  :)

**Comment by eskatos** [comment]
Our script template that the IDE loads and takes its parameters from specify that it should use 1.3 language and stdlib.

https://github.com/gradle/gradle/blob/67e966d162368f3b121b9f6cee213809e3d4b42d/subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/KotlinSettingsScript.kt#L71-L72

I guess the IDE is the one doing something wrong here given it works from the command line.

Could you please report the issue at https://kotl.in/issue ?

**Comment by oh-Rin** [comment]
In fact, I haven’t created any issue in [youtrack.jetbrains.com](https://youtrack.jetbrains.com/issues), it's a bit troublesome, but I think I’ll try it...

**Comment by RinOrz** [comment]
@eskatos hi, I have updated the gradle version to https://services.gradle.org/distributions-snapshots/gradle-7.0-20201227230035+0000-all.zip, but unfortunately this problem still exists, can this be solved?
![image](https://user-images.githubusercontent.com/58068445/103437642-d0e45e80-4c64-11eb-9500-eeb0a07a7c1c.png)

**Comment by bamboo** [comment]
> I have a plugin compiled with Kotlin 1.4.20 and referenced it in settings.gradle.kts, and then this error occurs. How can I solve it?

How are you compiling your plugin? Can you share all the Kotlin compiler options you're compiling it with?

**Comment by RinOrz** [comment]
> > I have a plugin compiled with Kotlin 1.4.20 and referenced it in settings.gradle.kts, and then this error occurs. How can I solve it?
> 
> How are you compiling your plugin? Can you share all the Kotlin compiler options you're compiling it with?

```kotlin
freeCompilerArgs = arrayOf(
  "-progressive",
  "-Xinline-classes",
  "-Xjvm-default=enable",
  "-Xallow-jvm-ir-dependencies",
  "-Xskip-prerelease-check",
  "-Xopt-in=kotlin.RequiresOptIn",
  "-Xopt-in=kotlin.time.ExperimentalTime",
  "-Xuse-experimental=kotlin.Experimental",
  "-Xuse-experimental=kotlin.ExperimentalStdlibApi",
  "-Xuse-experimental=kotlin.ExperimentalUnsignedTypes",
  "-Xuse-experimental=kotlin.contracts.ExperimentalContracts",
  "-Xuse-experimental=kotlinx.coroutines.ExperimentalCoroutinesApi"
)
```

**Comment by RinOrz** [comment]
I want to know if I am the only one with this problem? Seriously, this is annoying, because the kotlin version of all my current projects has been updated to 1.4.21. After compiling the plugin and using it in kts, this error will appear. It is really really annoying! I hope it can be resolved

**Comment by PHPirates** [comment]
I'm having the same problem, Kotlin 1.4.30-M1 and Gradle 6.8.3. Gradle clean does not help. Project files are here: https://github.com/Hannah-Sten/TeXiFy-IDEA/tree/install-without-restart

**Comment by petertrr** [comment]
I have faced this problem when adding a plugin as a dependency in `buildSrc`. With gradle **6.9**, for a plugin, compiled with Kotlin 1.5 with `useIR = true`, I get the same error. Changing kotlin compilation options in `buildSrc/build.gradle.kts` to
```kotlin
kotlinOptions {
    freeCompilerArgs = freeCompilerArgs + "-Xallow-jvm-ir-dependencies"
}
```
fixes these errors and everything seems to work fine. Is this intended behavior? Is it correct/safe to change compilation options in `buildSrc`?

**Comment by eskatos** [comment]
It was working as expected.
Gradle used the old backend and mixing dependencies built with the new backend caused that problem.

Starting with Gralde 8.0 that includes Kotlin 1.8.0 script compilation uses IR by default.
This issue is now fixed, closing.

For previous Gradle versions, the workaround shared above works:
* https://github.com/gradle/gradle/issues/15335#issuecomment-851943381
